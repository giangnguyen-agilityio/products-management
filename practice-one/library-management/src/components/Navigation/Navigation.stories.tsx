import { type StoryFn, type Meta } from '@storybook/react'
import { MemoryRouter } from 'react-router-dom'
import Navigation from './index'

export default {
  title: 'Example/Navigation',
  component: Navigation
} as Meta<typeof Navigation>

const Template: StoryFn<typeof Navigation> = (args) => (
  <MemoryRouter>
    <Navigation {...args} />
  </MemoryRouter>
)

export const Default = Template.bind({})
Default.args = {
  links: [
    { id: 'book', label: 'Book', url: '/book' },
    { id: 'member', label: 'Member', url: '/member' },
    { id: 'hire-request', label: 'Hire Request', url: '/request' }
  ]
}
