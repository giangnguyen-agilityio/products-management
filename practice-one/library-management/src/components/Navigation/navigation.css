.navigation {
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: var(--secondary-font);
  gap: 50px;
  padding: 20px;
  cursor: pointer;
}

.link-label {
  font-size: 26px;
  color: var(--primary-color);
  text-transform: uppercase;
  transition: all 0.2s ease-in-out;
}

.link-label:hover {
  padding-bottom: 10px;
  border-bottom: 1px solid var(--primary-color);
}
