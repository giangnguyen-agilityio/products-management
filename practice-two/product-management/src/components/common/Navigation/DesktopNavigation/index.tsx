import React from 'react'
import { Flex, Button } from '@chakra-ui/react'
import { Link } from 'react-router-dom'
import { NavigationLink } from '@types'

// Desktop Navigation Component
const DesktopNavigation: React.FC<{ links: NavigationLink[] }> = ({
  links,
}) => (
  <Flex as="nav" className="navigation" paddingTop={1.5}>
    <Flex as="ul" className="nav-list" gap={6}>
      {links.map(({ label, href }) => (
        <Flex as="li" className="nav-item" key={label}>
          <Button
            as={Link}
            to={href}
            aria-label={label}
            width="full"
            fontSize="md"
            textTransform="uppercase"
            color="textSecondary"
            variant="contract"
          >
            {label}
          </Button>
        </Flex>
      ))}
    </Flex>
  </Flex>
)

export default DesktopNavigation
