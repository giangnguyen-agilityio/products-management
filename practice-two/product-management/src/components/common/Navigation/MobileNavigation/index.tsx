// Libraries
import React from 'react'
import {
  Flex,
  Button,
  Drawer,
  DrawerBody,
  DrawerContent,
  DrawerOverlay,
  DrawerCloseButton,
} from '@chakra-ui/react'
import { Link } from 'react-router-dom'

// Types
import { NavigationLink } from '@types'

// Mobile Content Component
const MobileContent: React.FC<{
  isOpen: boolean
  toggleDrawer: () => void
  links: NavigationLink[]
}> = ({ isOpen, toggleDrawer, links }) => (
  <Drawer
    isOpen={isOpen}
    placement="left"
    onClose={toggleDrawer}
    size={{ base: 'full', sm: 'xs' }}
  >
    <DrawerOverlay />
    <DrawerContent className="sidebar" bgColor="black" opacity="0.95">
      {/* Close button for the mobile drawer */}
      <DrawerCloseButton
        width="40px"
        height="40px"
        fontSize="tiny"
        color="textTertiary"
        _hover={{
          color: 'textSecondary',
        }}
      />
      {/* Content of the mobile drawer */}
      <DrawerBody className="navigation-mobile">
        {/* Render mobile navigation links */}
        <MobileNavigationLinks links={links} />
      </DrawerBody>
    </DrawerContent>
  </Drawer>
)

// Mobile Navigation Links Component
const MobileNavigationLinks: React.FC<{ links: NavigationLink[] }> = ({
  links,
}) => (
  <Flex
    as="ul"
    className="nav-list-mobile"
    flexDir="column"
    marginTop={8}
    alignItems="center"
  >
    {/* Map through navigation links and render buttons */}
    {links.map(({ label, href }) => (
      <Flex as="li" className="nav-item-mobile" key={label}>
        <Button
          as={Link}
          to={href}
          display="flex"
          aria-label={label}
          width="full"
          height="40px"
          fontSize="small"
          variant="navigate"
        >
          {label}
        </Button>
      </Flex>
    ))}
  </Flex>
)

export default MobileContent
