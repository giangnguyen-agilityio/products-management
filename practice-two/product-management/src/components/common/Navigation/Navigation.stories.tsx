import type { Meta, StoryObj } from '@storybook/react'
import Navigation from './index'

const meta = {
  title: 'Practice Two/Navigation Component',
  component: Navigation,
  tags: ['autodocs'],
} satisfies Meta<typeof Navigation>

export default meta
type Story = StoryObj<typeof meta>

export const Primary: Story = {
  args: {
    links: [
      { label: 'home', href: 'javascript:void(0)' },
      { label: 'about us', href: 'javascript:void(0)' },
      { label: 'contact', href: 'javascript:void(0)' },
    ],
  },
}
