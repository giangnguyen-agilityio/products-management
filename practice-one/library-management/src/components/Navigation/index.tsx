import './navigation.css'
import { Link } from 'react-router-dom'

interface NavLink {
  id: string
  label: string
  url: string
}

interface NavigationProps {
  links: NavLink[]
}

const Navigation = (props: NavigationProps) => {
  const { links } = props
  return (
    <ul className="navigation">
      {links.map((link) => (
        <li key={link.id} id={`${link.id}-page`}>
          <Link to={link.url} className="link-label">
            {link.label}
          </Link>
        </li>
      ))}
    </ul>
  )
}

export default Navigation
