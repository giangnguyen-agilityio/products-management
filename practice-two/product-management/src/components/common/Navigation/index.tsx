// Libraries
import React, { memo } from 'react'
import { useDisclosure, Show, Hide, IconButton } from '@chakra-ui/react'

// Components
import DesktopNavigation from './DesktopNavigation'
import MobileContent from './MobileNavigation'

// Assets
import { HamburgerIcon } from '@chakra-ui/icons'

// Types
import { NavigationLink } from '@types'

interface NavigationProps {
  links: NavigationLink[]
}

const Navigation: React.FC<NavigationProps> = ({ links }) => {
  const { isOpen, onToggle } = useDisclosure()

  return (
    <>
      {/* Render the Desktop Navigation above medium screen size */}
      <Show above="md">
        <DesktopNavigation links={links} />
      </Show>

      {/* Render the Hamburger Menu button for smaller screens */}
      <Hide above="md">
        <IconButton
          className="hamburger-menu"
          aria-label="Open Menu"
          fontSize="mediumLarge"
          paddingTop={1}
          icon={<HamburgerIcon />}
          onClick={onToggle}
          variant="unstyled"
          color="textSecondary"
        />
      </Hide>

      {/* Render the Mobile Navigation content */}
      <MobileContent isOpen={isOpen} toggleDrawer={onToggle} links={links} />
    </>
  )
}

export default memo(Navigation)
