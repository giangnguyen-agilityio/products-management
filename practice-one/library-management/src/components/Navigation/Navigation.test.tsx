import React from 'react'
import {render} from '@testing-library/react'
import {MemoryRouter} from 'react-router-dom'
import * as helpers from '@helpers'
import Navigation from '.'

// Mocking getItemInLocalStorage function
jest.mock('@helpers', () => ({
  getItemInLocalStorage: jest.fn(() => 'memberId'),
}))

// Test data for navigation links
const testLinks = [
  {id: 'link1', label: 'Link 1', url: '/link1'},
  {id: 'link2', label: 'Link 2', url: '/link2'},
  {id: 'link3', label: 'Link 3', url: '/link3'},
]

describe('Navigation component', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should renders navigation links for a member', () => {
    const {container, getByText} = render(
      <MemoryRouter>
        <Navigation links={testLinks} />
      </MemoryRouter>
    )

    testLinks.forEach(({id, label}) => {
      const linkElement = getByText(label)
      expect(linkElement.tagName).toBe('A') // Should render as NavLink
      expect(linkElement).toHaveAttribute('href', `/${id}`)
    })
    expect(container).toMatchSnapshot()
  })

  it('should renders navigation links as span for non-member', () => {
    // Simulate the user is not a member
    jest.spyOn(helpers, 'getItemInLocalStorage').mockReturnValueOnce('')

    const {container, getByText} = render(
      <MemoryRouter>
        <Navigation links={testLinks} />
      </MemoryRouter>
    )

    testLinks.forEach(({label}) => {
      const linkElement = getByText(label)
      expect(linkElement.tagName).toBe('SPAN') // Should render as span
    })
    expect(container).toMatchSnapshot()
  })
})
