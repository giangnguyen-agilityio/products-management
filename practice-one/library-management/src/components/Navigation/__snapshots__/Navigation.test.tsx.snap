// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component should renders navigation links as span for non-member 1`] = `
<div>
  <ul
    class="navigation"
  >
    <li
      id="link1-page"
    >
      <span
        class="link-label"
      >
        Link 1
      </span>
    </li>
    <li
      id="link2-page"
    >
      <span
        class="link-label"
      >
        Link 2
      </span>
    </li>
    <li
      id="link3-page"
    >
      <span
        class="link-label"
      >
        Link 3
      </span>
    </li>
  </ul>
</div>
`;

exports[`Navigation component should renders navigation links for a member 1`] = `
<div>
  <ul
    class="navigation"
  >
    <li
      id="link1-page"
    >
      <a
        class="link-label"
        href="/link1"
      >
        Link 1
      </a>
    </li>
    <li
      id="link2-page"
    >
      <a
        class="link-label"
        href="/link2"
      >
        Link 2
      </a>
    </li>
    <li
      id="link3-page"
    >
      <a
        class="link-label"
        href="/link3"
      >
        Link 3
      </a>
    </li>
  </ul>
</div>
`;
